File: app/src/main/java/io/dume/dume/util/StateManager.java
package io.dume.dume.util;

import android.content.Context;
import android.content.SharedPreferences;

public class StateManager {

    public static StateManager instance;
    private static SharedPreferences local;
    private SharedPreferences.Editor edit;

    private StateManager() {
    }

    //get the singleton of the stateManager
    public static StateManager getInstance(Context context) {
        if (instance == null) {
            instance = new StateManager();
            local = context.getSharedPreferences("state", Context.MODE_PRIVATE);
        }
        return instance;
    }

    //set any value to the sharePreference
    public void setValue(String key,  Object  value) {
        edit = local.edit();
        if (value instanceof String) {
            edit.putString(key, (String) value);
        } else if (value instanceof Integer) {
            edit.putInt(key, (Integer) value);
        } else if (value instanceof Boolean) {
            edit.putBoolean(key, (Boolean) value);
        } else if (value instanceof Float) {
            edit.putFloat(key, (Float) value);
        }else if (value instanceof Long) {
            edit.putLong(key, (Long) value);
        }
        edit.apply();
    }

    //get replacement
    public SharedPreferences getLocal() {
        return local;
    }

    //current state in the forward flow
    public void setTeacherCurrentState(String state) {
        edit.putString("teacherCurrentState", state);
        edit.apply();
    }

    public void setStudentCurrentState(String state) {
        edit.putString("teacherCurrentState", state);
        edit.apply();
    }

    public void setRole(String role) {
        edit.putString("role", role);
        edit.apply();
    }

    public void setFirstTimeUser(Boolean state) {
        edit.putBoolean("firstTimeUser", state);
        edit.apply();
    }
}

File: app/src/main/java/io/dume/dume/util/StateManager.kt
package io.dume.dume.util

import android.content.Context
import android.content.SharedPreferences

class StateManager private constructor() {

    fun setValue(key: String, value: Any) {

        if (value is String) {
            edit!!.putString(key, value)
        } else if (value is Int) {
            edit!!.putInt(key, value)
        } else if (value is Boolean) {
            edit!!.putBoolean(key, value)
        } else if (value is Float) {
            edit!!.putFloat(key, value)
        } else if (value is Long) {
            edit!!.putLong(key, value)
        }
        edit!!.apply()
    }

    //get replacement
    fun sharedPreferences(): SharedPreferences {
        return sharedPreferences
    }

    //current state in the forward flow
    fun setTeacherCurrentState(state: String) {
        edit!!.putString(TEACHER_POSITION, state)
        edit!!.apply()
    }

    fun setStudentCurrentState(state: String) {
        edit!!.putString(STUDENT_POSITION, state)
        edit!!.apply()
    }

    fun setRole(role: String) {
        edit!!.putString(ROLE, role)
        edit!!.apply()
    }


    fun setFirstTimeUser(state: Boolean?) {
        edit!!.putBoolean(FIRST_TIME_USER, state!!)
        edit!!.apply()
    }


    companion object {

        lateinit var sharedPreferences: SharedPreferences
        private var edit: SharedPreferences.Editor? = null
        public val FIRST_TIME_USER = "firstTimeUser"
        public val TEACHER_POSITION = "teacherCurrentState"
        public val STUDENT_POSITION = "studentCurrentState"
        public val ROLE = "role"

        fun getInstance(context: Context): StateManager {
            sharedPreferences = context.getSharedPreferences("state", Context.MODE_PRIVATE)
            edit = sharedPreferences!!.edit()
            return StateManager()
        }
    }
}
